### PR TITLE
Feature/Polish after feedback

### PR DESCRIPTION
**Pull request by yarkhinephyo**
### Summary
Adopt feedback for minor UX edits and fix various bugs in the frontend before launch. No major feature changes.

#### After user review
- Add sharing of link in gym page and user page
- Show user statistics of upvotes and posts
- Add help button in `New` page
- Fix screenshot low resolution problem in `About` page
- Fix multiple calling of `onMounted` when refreshed by caching
- Add caching to `routesByUser`, `gymsByCountry`, `routesByGym`, `routeDetails`
- Reset canvas after uploading
- Remove loading overlay spinner
- Add terms and conditions and privacy page which might not be viewable in dev environment

#### After 1st PR review
- Add feedback button
- Sort alphabetically by gym name 
- Add color to vote and share buttons
- Prompts user to log in if tries to vote
- Change route name maxlength from 30 to 20 in frontend, backend

### CODE REVIEW

File: ionic_user_interface/src/common/shareSocial.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { Share } from '@capacitor/share';
+import { RouteLocationNormalizedLoaded } from 'vue-router';
+
+// Cannot use BASE_URL to get domain URL in native app
+// TODO: If native platform, return fixed domain. If web, return BASE_URL
+const DOMAIN = 'https://routemaker.rocks';
+
+const shareSocial = async (vueRoute: RouteLocationNormalizedLoaded, message: string) => {
+  return Share.share({
```
Comment (qwoprocks) [inline ionic_user_interface/src/common/shareSocial.ts:9]: Apparently this doesn't work in firefox browser... :(

File: ionic_user_interface/src/common/api/route/getRoutesByUser.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    return getRoutesByUser(username);
+  });
+};
+
+interface ResponseData {
+  Message: string;
+  Items: UserRoute[];
+}
+
+interface UserRoute {
```
Comment (qwoprocks) [inline ionic_user_interface/src/common/api/route/getRoutesByUser.ts:37]: Actually, this looks to be a superset of GymRoutes, maybe we can subclass? Up to you tho

File: ionic_user_interface/src/common/api/route/getRoutesByGym.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import axios from 'axios';
```
Comment (qwoprocks) [inline ionic_user_interface/src/common/api/route/getRoutesByGym.ts:1]: Nice refactoring 👍

File: ionic_user_interface/src/views/Signup.vue [hunk 4]
```diff
@@ -70,6 +70,19 @@
                 >
                   Sign up
                 </ion-button>
+                <div class="privacy-policy">
+                  <ion-checkbox v-model="termsCheckbox" />
+                  <ion-text>
+                    I agree to
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Signup.vue:76]:
```suggestion
                    I agree to the
```

File: ionic_user_interface/src/views/Signup.vue [hunk 5]
```diff
@@ -70,6 +70,19 @@
...
+                    <a href="https://routemaker.rocks/privacy.html" target="_blank">
+                      Privacy Policy
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Signup.vue:78]: image

The underline is a little weird, maybe you can try changing `&#38;` to `&nbsp;&#38;&nbsp;`

File: ionic_user_interface/src/views/Signup.vue [hunk 6]
```diff
@@ -275,4 +296,22 @@ export default defineComponent({
...
   margin-bottom: 40px;
 }
+
+.privacy-policy ion-text {
+  font-size: 0.7em;
+  vertical-align: middle;
+  display: inline-block;
+  margin-left: 0.5em;
+
+  position: relative;
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Signup.vue:306]: Can try using display flex instead for vertical alignment.

File: ionic_user_interface/src/views/New.vue [hunk 7]
```diff
@@ -16,15 +16,23 @@
           :width="canvasWidth"
           :ionContent="ionContent"
         />
+        <div v-if="!photoUploaded">
+          <br />
+          <ion-button @click="handleClickUserGuide" color="medium" fill="clear">
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/New.vue:21]: Idk if it's just me but I feel like adding this button makes the page look off-balanced (not horizontally centered) or smth

File: ionic_user_interface/src/views/Help.vue [hunk 8]
```diff
@@ -107,16 +112,18 @@ export default defineComponent({
 
 .screenshot {
   border: 1.5px solid grey;
-  max-width: 200px;
+  max-width: 250px;
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Help.vue:115]: Still feels a little small on desktop, maybe we can do like 300px?

File: ionic_user_interface/src/views/Help.vue [hunk 9]
```diff
@@ -19,13 +19,17 @@
           </p>
         </div>
 
-        <p class="lead">Please star our repository to support!</p>
+        <p class="lead">
+          We plan to maintain this project for as long as people find it useful.
+          <br />
+          Please star our repository to support. Pull requests are welcomed too!
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Help.vue:25]:
```suggestion
          If you like our work, please star our repository. Pull requests are welcomed too!
```

File: ionic_user_interface/src/common/api/route/getRoutesByUser.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import axios from 'axios';
+import { routeBaseUrl } from './config';
```
Comment (qwoprocks) [inline ionic_user_interface/src/common/api/route/getRoutesByUser.ts:2]: This and the config file seems a little unnecessary lol

File: ionic_user_interface/src/components/gym-selector/GymRequestForm.vue [hunk 11]
```diff
@@ -97,7 +97,7 @@ export default defineComponent({
     const alertSubmissionSuccess = async () => {
       const alert = await alertController.create({
         header: 'Request Submitted',
-        subHeader: 'Thank you! We will respond to you in 12 hours',
+        subHeader: 'Thank you! A response email will be sent in 3 to 6 hours',
```
Comment (qwoprocks) [inline ionic_user_interface/src/components/gym-selector/GymRequestForm.vue:100]: Same comment, seems a little too short of a time frame. I suggest 2-12 hours. Also, are we gonna manually email them back?

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: ionic_user_interface/src/views/Signup.vue [hunk 12]
```diff
@@ -70,6 +70,19 @@
                 >
                   Sign up
                 </ion-button>
+                <div class="privacy-policy">
+                  <ion-checkbox v-model="termsCheckbox" />
+                  <ion-text>
+                    I agree to
+                    <a href="https://routemaker.rocks/privacy.html" target="_blank">
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Signup.vue:77]: Maybe just href="/privacy.html"? Also, we might wanna consider putting this as an inline scrollable component like other websites

File: ionic_user_interface/src/views/Signup.vue [hunk 13]
```diff
@@ -70,6 +70,19 @@
                 >
                   Sign up
                 </ion-button>
+                <div class="privacy-policy">
+                  <ion-checkbox v-model="termsCheckbox" />
+                  <ion-text>
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/Signup.vue:75]: image

Need to import.

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: ionic_user_interface/src/views/ViewRoute.vue [hunk 14]
```diff
@@ -8,6 +8,9 @@
           <div class="route-title">
             <b>{{ routeDetails.routeName }}</b>
           </div>
+          <div class="share-button-div icon-button margin-right" @click="sharePostHandler">
+            <ion-icon size="large" :icon="shareSocialOutline"></ion-icon>
+          </div>
           <VoteButton
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/ViewRoute.vue:14]: Oops, I think the votebutton doesn't have `cursor: pointer` on hover currently, can help me to add in?

File: ionic_user_interface/src/common/api/route/getGymsByCountry.ts [hunk 15]
```diff
@@ -1,5 +1,12 @@
 import axios from 'axios';
 import { routeBaseUrl } from './config';
+import cacheManager from 'cache-manager';
+
+const memoryCache = cacheManager.caching({
+  store: 'memory',
+  max: 10, // Number of items in cache
```
[the remark this answers is not in the document]
Comment (yarkhinephyo) [reply ionic_user_interface/src/common/api/route/getGymsByCountry.ts:7]: If there are alrdy 10 items in cache, the 11th item will pop the 1st item. TTL means the number of seconds the item will stay in cache.

File: ionic_user_interface/src/views/RequestGym.vue [hunk 16]
```diff
@@ -5,6 +5,10 @@
         <strong>Route Maker</strong>
         <p>Climbing Gym Request Form</p>
         <br />
+        <p class="description">
+          Please submit gym requests here. We will respond to you via email in 3 to 6 hours. Thank
```
[the remark this answers is not in the document]
Comment (yarkhinephyo) [reply ionic_user_interface/src/views/RequestGym.vue:9]: Yea we will. Lemme think manual or auto

File: ionic_user_interface/src/views/Signup.vue (hunk 12 above)
Comment (yarkhinephyo) [reply ionic_user_interface/src/views/Signup.vue:77]: If we change to native, it won't work anymore right? Cos native no domain

File: ionic_user_interface/src/views/ViewRoute.vue (hunk 14 above)
Comment (yarkhinephyo) [reply ionic_user_interface/src/views/ViewRoute.vue:14]: U added in the javascript

```
const css = '.vote-button-component:hover { background-color: #333333; cursor: pointer; }';
```

File: ionic_user_interface/src/common/api/route/getRoutesByUser.ts (hunk 10 above)
Comment (yarkhinephyo) [reply ionic_user_interface/src/common/api/route/getRoutesByUser.ts:2]: This was in case we hav many microservices then in the end decide to map them all to domain for example `routemaker.rocks/api`. Yea I ll just remove it since we are not doing that

File: ionic_user_interface/src/views/New.vue (hunk 7 above)
Comment (yarkhinephyo) [reply ionic_user_interface/src/views/New.vue:21]: Removed the icon to center the text

File: ionic_user_interface/src/views/Signup.vue (hunk 12 above)
Comment (qwoprocks) [reply ionic_user_interface/src/views/Signup.vue:77]: Ah right, ok I think maybe we should just add the terms and condition as an inline scrollable div or smth

File: ionic_user_interface/src/common/api/route/getGymsByCountry.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (qwoprocks) [reply ionic_user_interface/src/common/api/route/getGymsByCountry.ts:6]: Why do we need like more than 1 item in cache though? I thought we just need the latest version?

File: ionic_user_interface/src/components/VoteButton.vue [hunk 17]
```diff
@@ -117,13 +128,19 @@ export default defineComponent({
...
   font-size: 16px;
   border-radius: 10px;
-  border: 2px solid grey;
+  border: 2px solid var(--ion-color-danger);
   padding: 5px 7px;
   align-self: center;
 }
 
+.vote-button-component:hover {
+  background-color: var(--ion-color-danger-tint);
```
Comment (qwoprocks) [inline ionic_user_interface/src/components/VoteButton.vue:137]: image
The tint is too light and blends in with the bg, maybe try a slightly darker color? Alternatively only tint the heart itself

File: ionic_user_interface/src/views/UserRoutes.vue [hunk 18]
```diff
@@ -5,41 +5,55 @@
...
           <ion-col class="ion-align-self-center" size-lg="6" size-md="8" size-xs="12">
-            <ion-button
-              v-if="isAdmin"
-              class="margin-left"
-              color="danger"
-              @click="disableUserHandler"
-            >
-              <ion-label>Admin Disable</ion-label>
-            </ion-button>
+            <ion-row class="ion-justify-content-center">
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/UserRoutes.vue:31]: This row/col is not aligned with the other stuff on the page
image

File: ionic_user_interface/src/views/RequestGym.vue [hunk 19]
```diff
@@ -5,6 +5,10 @@
         <strong>Route Maker</strong>
         <p>Climbing Gym Request Form</p>
         <br />
+        <p class="description">
```
Comment (qwoprocks) [inline ionic_user_interface/src/views/RequestGym.vue:8]: Font-color is black even in dark mode, cannot read at all

File: ionic_user_interface/src/components/GymRouteList.vue [hunk 20]
```diff
@@ -180,6 +168,10 @@ enum SortMode {
   NEWEST,
 }
 
+interface GymRouteWithId extends GymRoute {
+  routeId?: number;
```
Comment (qwoprocks) [inline ionic_user_interface/src/components/GymRouteList.vue:172]: Does this actually need the `?` ?

**Review by qwoprocks — COMMENTED** [review]
Some comments

File: ionic_user_interface/src/views/RequestGym.vue (hunk 19 above)
Comment (yarkhinephyo) [reply ionic_user_interface/src/views/RequestGym.vue:8]: good catch

File: ionic_user_interface/src/components/VoteButton.vue [hunk 21]
```diff
@@ -52,16 +52,27 @@ export default defineComponent({
 
     let waitingForResponse = false;
 
-    // Display hover feedback only if logged in
```
Comment (qwoprocks) [inline ionic_user_interface/src/components/VoteButton.vue:55]: Hmmm why is this removed?

File: ionic_user_interface/src/components/VoteButton.vue [hunk 22]
```diff
@@ -117,13 +128,19 @@ export default defineComponent({
...
   font-size: 16px;
   border-radius: 10px;
-  border: 2px solid grey;
+  border: 2px solid var(--ion-color-danger);
   padding: 5px 7px;
   align-self: center;
 }
 
+.vote-button-component:hover {
+  background-color: var(--ion-color-danger-shade);
```
Comment (qwoprocks) [inline ionic_user_interface/src/components/VoteButton.vue:137]: image
Hover bgColor is still blending in with the text/outline

**Review by qwoprocks — APPROVED** [review]

File: ionic_user_interface/src/components/VoteButton.vue (hunk 21 above)
Comment (yarkhinephyo) [reply ionic_user_interface/src/components/VoteButton.vue:55]: Cos if he is logged in, he can still get hover feedback and also click. But there will be toast telling him to log in. This should encourage people to sign up instead of thinking the vote button got problem

Comment (qwoprocks) [reply ionic_user_interface/src/components/VoteButton.vue:55]: Ok, makes sense
